extension/src/openvic-extension/singletons/CursorSingleton.cpp

```diff
@@ -0,0 +1,622 @@
+	static const String backslash = "\\";
+	static const String forwardslash = "/";
+	static const String cursor_directory_forwardslash = Utilities::std_to_godot_string(cursor_directory) + forwardslash;
+	static const String dot = ".";
+	return path.replace(backslash, forwardslash).get_slice(cursor_directory_forwardslash, 1).get_slice(dot, 0);
+}
+
+bool CursorSingleton::load_cursors() {
+	GameSingleton const* game_singleton = GameSingleton::get_singleton();
+	ERR_FAIL_NULL_V_MSG(game_singleton, false, "Error retrieving GameSingleton");
+	
+	
+
+	//there is also a png file in the folder we don't want to bother loading
+	//so don't just load everything in the directory
+
+	//We need to load both ".cur" and ".CUR" files
+	Dataloader::path_vector_t cursor_files = game_singleton->get_dataloader()
+		.lookup_files_in_dir_recursive(cursor_directory, ".cur");
+
+	Dataloader::path_vector_t CURsor_files = game_singleton->get_dataloader()
+		.lookup_files_in_dir_recursive(cursor_directory, ".CUR");
+	cursor_files.insert(std::end(cursor_files),std::begin(CURsor_files),std::end(CURsor_files));
+
+	Dataloader::path_vector_t animated_cursor_files = game_singleton->get_dataloader()
+		.lookup_files_in_dir_recursive(cursor_directory, ".ani");
+	
+	if (cursor_files.empty() && animated_cursor_files.empty()){
+		Logger::error("failed to load cursors: no files in cursors directory");
+		return false;
+	}
+
+	bool ret = true;
+
+	for(fs::path const& file_name : cursor_files) {
+		String file = Utilities::std_to_godot_string(file_name.string());
+		StringName name = to_define_file_name(file);
+
+		if (!_load_cursor_cur(name,file)){
+			Logger::error("failed to load normal cursor at path ", file_name);
+			ret = false;
+		}
+	}
+
+	for(fs::path const& file_name : animated_cursor_files) {
+		String file = Utilities::std_to_godot_string(file_name.string());
+		StringName name = to_define_file_name(file);
+
+		if (!_load_cursor_ani(name,file)){
+			Logger::error("failed to load animated cursor at path ", file_name);
+			ret = false;
+		}
+	}
+
+	return ret;
+}
+
+static int32_t _reverse_bits(int32_t byte, int32_t bitsPerPixel=8) {
+	int32_t reverser_lookup[] = {
+		0x0, 0x8, 0x4, 0xc, 0x2, 0xa, 0x6, 0xe,
+		0x1, 0x9, 0x5, 0xd, 0x3, 0xb, 0x7, 0xf
+	};
+	int32_t a = reverser_lookup[(byte & 0b1111)] << 4;
+	int32_t b = reverser_lookup[byte >> 4];
+	int32_t c = b | a;
+	return c >> (8-bitsPerPixel);
+}
+
+static int32_t _rotate_right(int32_t byte, int32_t size=8) {
+	return ((byte & 0b1) << (size-1)) | (byte >> 1);
+}
+
+static int32_t _load_int_256(Ref<FileAccess> const& file) {
+	int32_t value = file->get_8();
+	if (value == 0) value = 256;
+	return value;
+}
+
+static int32_t _get_row_start(Vector2i dimensions, int32_t y_coord, int32_t bitsPerPixel) {
+	int32_t rowCount = dimensions.x * bitsPerPixel / 32;
+	bool hasExtraRow = ((dimensions.x * bitsPerPixel) % 32) != 0;
+	rowCount += hasExtraRow; //TODO 1*
+	return rowCount*y_coord*4; //4 bytes per row * rows down
+}
```

Comment:
```suggestion
static constexpr int32_t _get_row_start(int32_t x_coord, int32_t y_coord, int32_t bitsPerPixel) {
	x_coord *= bitsPerPixel;
	int32_t rowCount = (x_coord + 31) >> 5;
	return rowCount * y_coord * 4; // 4 bytes per row * rows down
}
```
- On further inspection, this function can also be `constexpr` if we just replace the `Vector2i` with an `int32_t` (since we only use its x coordinate anyway) and then just use `dimensions.x` as its first argument later instead of  `dimensions`.
- What you're doing with `rowCount` and `hasExtraRow` is a common operation with a classic solution (and further, a specialisation for cases like this one involving power of 2):
  - If we separate out the `* bitsPerPixel` and just apply that to `x_coord` at the beginning, we're left with an integer divide and round up.
  - The classic solution to this problem is `(x + divisor - 1) / divisor`, where the `+ divisor - 1` ensures the result will be 1 greater if `x % divisor != 0`, but has no effect if `x` is an exact multiple of `divisor`.
  - With powers of two we can further specialise this by replacing the `/ divisor` with `>> log2(divisor)`, in this case `>> 5`